### PR TITLE
YARD completion snippet after typing `##`

### PR DESCRIPTION
**Pull request by jez**
### Motivation

Hoping to encourage more people to write documentation comments in our codebase.

Fixes https://github.com/sorbet/sorbet/issues/5402

### Test plan

See included automated tests.

### CODE REVIEW

**Comment by jez** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_LOBTBknX0GuNL2
→ https://go/builds/bui_LOBTsKSYEKYvRJ

**Comment by jez** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_LOBoAzciK73sp0
→ https://go/builds/bui_LOBoIb6tCCAsWZ

File: test/helpers/lsp.cc [hunk 1]
```diff
@@ -508,6 +510,19 @@ vector<unique_ptr<LSPMessage>> getLSPResponsesFor(LSPWrapper &wrapper, unique_pt
     return getLSPResponsesFor(wrapper, move(messages));
 }
 
+namespace {
+
+// Like absl::StripTrailingAsciiWhitespace, but only blank characters (tabs and spaces)
+ABSL_MUST_USE_RESULT inline absl::string_view stripTrailingAsciiBlank(absl::string_view str) {
```
Comment (elliottt) [inline test/helpers/lsp.cc:516]: Does `[[nodiscard]]` not work here? It looks like abseil suggests using that instead for c++17 and above: https://github.com/abseil/abseil-cpp/blob/0c6302fe427963ec5c471d3ee660120682ab15f7/absl/base/attributes.h#L403-L404

```suggestion
[[nodiscard]] inline absl::string_view stripTrailingAsciiBlank(absl::string_view str) {
```

File: main/lsp/requests/completion.cc [hunk 2]
```diff
@@ -681,28 +681,6 @@ unique_ptr<CompletionItem> trySuggestSig(LSPTypecheckerInterface &typechecker,
...
-    if (receiverSym == core::Symbols::rootSingleton()) {
-        receiverSym = core::Symbols::Object().data(gs)->lookupSingletonClass(gs);
-    }
-    auto methodOwner = targetMethod.data(gs)->owner;
-
-    if (!(methodOwner == receiverSym || methodOwner == receiverSym.data(gs)->attachedClass(gs))) {
-        // The targetMethod we were going to suggest a sig for is not actually in the same scope as this sig.
-        return nullptr;
-    }
-
```
Comment (elliottt) [inline main/lsp/requests/completion.cc:705]: Is this all redundant after the changes to `NextMethodFinder`?

**Review by elliottt — APPROVED** [review]
This looks good!

File: test/helpers/lsp.cc (hunk 1 above)
Comment (jez) [reply test/helpers/lsp.cc:516]: It probably works, I just copied the implementation of `StripTrailingAsciiWhitespace` verbatim with no changes other than the `is_blank` thing.

I'll change this in a follow up so i don't have to wait for tests to build for this PR.

File: main/lsp/requests/completion.cc (hunk 2 above)
Comment (jez) [reply main/lsp/requests/completion.cc:705]: Yes. I had to do this in NextMethodFinder for two reasons:

- I wanted to share this logic between suggested sigs and suggested YARD docs
- This logic relies on looking at the `enclosingMethod` that comes back in the LSP QueryResponse to figure out whether the method we find is owned by the same owner as the `enclosingMethod.data(gs)->owner`

For the YARD snippet case, we don't have any QueryResponses, which means no enclosing method, which means we need a purely syntactic way to check that we're in the same nesting scope as the queryLoc.

**Pull request merged**